glogger/handler.py

```diff
@@ -0,0 +1,277 @@
+        self.flush_threshold = flush_threshold  # force flush if buffer size reaches this percentage of capacity
+        self.max_send_tries = max_send_tries  # maximum number of times to retry sending logs if request fails
+        self.messages_buffer = MessagesBuffer(max_total_length, overflow_drop_factor)
+        self.messages_buffer.head_serial_no = continue_from
+        self.stop_event = threading.Event()
+        self.time_fn = time.time
+        self.jsonify = JSONEncoder(separators=(",", ":")).encode  # compact, no whitespace
+        self.uri = f"{self.scheme}://{server_address}/api/v1/logs"
+        self.session = requests.Session()
+        self.session.headers.update(
+            {
+                "X-Application-Name": application_name,
+                "X-Token": auth_token,
+            }
+        )
+        self.last_flush_time = 0.0
+        self.flush_thread = threading.Thread(target=self._flush_loop, daemon=True, name="Logs flusher")
+        self.start()
+
+    def start(self) -> None:
+        """Called to start the flush thread."""
+        assert not self.stop_event.is_set(), "restart not allowed!"
+        self.flush_thread.start()
+
+    def emit(self, record: LogRecord) -> None:
+        # Called while lock is acquired
+        prev_dropped = self.messages_buffer.dropped
+        self.messages_buffer.append(self.format(record))
+        new_dropped = self.messages_buffer.dropped - prev_dropped
+        if new_dropped:
+            self.warn_dropped(new_dropped)
+
+    def format(self, record: LogRecord) -> str:
+        super().format(record)
+        formatted_timestamp = datetime.utcfromtimestamp(record.created).isoformat()
+        d = {
+            **self.get_extra_fields(record),
+            "message": record.message,
+            "serial_no": self.messages_buffer.next_serial_no,
+            "severity": record.levelno,
+            "timestamp": formatted_timestamp,
+            "logger_name": record.name,
+        }
+        s = self.jsonify(d)
+        if len(s) > self.max_message_size:
+            self.truncate(d)
+            s = self.jsonify(d)
+            assert len(s) <= self.max_message_size, "did not truncate enough!"
+        return s
+
+    def get_extra_fields(self, record: LogRecord) -> dict:
+        """Override to add extra fields to formatted record."""
+        return {}
+
+    def truncate(self, d):
+        minimum = 80
+        long_items = [(k, v) for k, v in d.items() if isinstance(v, str) and len(v) > minimum]
+        # Leave a KB buffer for all the short items + any overhead:
+        max_field_length = max(minimum, (self.max_message_size - 1024) // len(long_items))
+        for k, v in long_items:
+            d[k] = v[:max_field_length]
+        d["truncated"] = True
+
+    def warn_dropped(self, dropped: int) -> None:
+        # If we use logging, we risk getting here more than once. No can do.
```

Comment:
I really don't like this, much copied code and format will need to be changed in other places.
Would pass to server in next message as its own field (and zero the number of dropped only after we sent successfully)

This will require slight change in server.